### PR TITLE
Avoid uninitialized NodeId, SubmapId

### PR DESCRIPTION
**Pull request by gaschler**
Previously, NodeId and SubmapId could be (partially) uninitialized,
for instance like this:

```
NodeId node_id;
SubmapId submap{0}; // uninitialized submap_index
```

This introduces constructors to prevent this.

### CODE REVIEW

**Comment by gaschler** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user gaschler. Merge queue now has a length of 2.

File: cartographer/mapping/internal/constraints/constraint_builder_2d_test.cc [hunk 1]
```diff
@@ -79,12 +79,12 @@ TEST_F(ConstraintBuilder2DTest, FindsConstraints) {
   for (int i = 0; i < 2; ++i) {
     EXPECT_EQ(constraint_builder_->GetNumFinishedNodes(), expected_nodes);
     for (int j = 0; j < 2; ++j) {
-      constraint_builder_->MaybeAddConstraint(submap_id, &submap, NodeId{},
+      constraint_builder_->MaybeAddConstraint(submap_id, &submap, NodeId{0, 0},
```
Comment (pifon2a) [inline cartographer/mapping/internal/constraints/constraint_builder_2d_test.cc]: Maybe just specify defaults in `NodeId` struct directly?

Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d_test.cc]: In my eyes, a Node/SubmapId should never be default-initialized (outside of tests).
If you don't set it, you shouldn't create an id.

Comment (pifon2a) [reply cartographer/mapping/internal/constraints/constraint_builder_2d_test.cc]: Then maybe delete the default constructor?

**Pull request merged**